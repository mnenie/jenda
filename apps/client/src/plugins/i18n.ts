import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
