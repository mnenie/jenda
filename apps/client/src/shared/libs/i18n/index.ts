import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
