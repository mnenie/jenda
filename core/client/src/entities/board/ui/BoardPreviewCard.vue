<script setup lang="ts">
import type { User } from '@/entities/user/model'
import type { BoardPreview } from '../model'
import { UserAvatar } from '@/entities/user'
import { computed } from 'vue'

defineProps<{
  board: BoardPreview
}>()

const userPosition = computed(() => {
  return (board: BoardPreview, user: User) => {
    return (board.users.length - 1 - +user._id!) * 12
  }
})
</script>

<template>
  <div :class="$style.active_board" @click="$router.push('/board/1')">
    <div :class="$style.text">
      <p class="text-base" style="font-weight: 500">
        {{ board.name }}
      </p>
      <span class="text-xs">{{ board.description }}</span>
    </div>
    <div :class="$style.bottom_part">
      <span class="text-xs"> {{ $t('boards.card.date_updated') }}: May 2024</span>
      <div :class="$style.users">
        <template v-for="user in board.users" :key="user._id">
          <UserAvatar
            :class="$style.user"
            :style="{
              right: `${userPosition(board, user)}px`,
              zIndex: userPosition(board, user),
              border: `1px solid ${user._id === '0' ? 'var(--purple-main)' : 'var(--zinc-300)'}`,
            }"
          >
            <img :src="user.photoUrl" style="width: 100%" />
          </UserAvatar>
        </template>
      </div>
    </div>
  </div>
</template>

<style module lang="scss">
.active_board {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: white;
  border: 1px solid var(--zinc-200);
  min-width: 240px;
  height: 120px;
  border-radius: 15px;
  padding: 15px;
  cursor: pointer;

  .text {
    & span {
      color: var(--zinc-500);
    }
  }

  .bottom_part {
    display: flex;
    align-items: flex-end;
    justify-content: space-between;
    width: 100%;

    & > span {
      color: var(--zinc-500);
    }

    .users {
      position: relative;
      display: flex;
      align-items: center;
      height: 20px;
      width: 20px;

      .user {
        height: 22px;
        width: 22px;
        position: absolute;
        top: 0;
      }
    }
  }
}

:global(html.dark) {
  .active_board {
    background-color: rgba(var(--zinc-rgb-600), 0.4);
    border-color: var(--zinc-600);
    color: var(--zinc-100);

    .text {
      & p {
        color: var(--zinc-100);
      }
      & span {
        color: var(--zinc-200);
      }
    }

    .bottom_part {
      & > span {
        color: var(--zinc-300);
      }
    }
  }
}
</style>
