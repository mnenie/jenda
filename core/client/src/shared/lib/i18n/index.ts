import { useCookies } from '@vueuse/integrations/useCookies'
import { computed } from 'vue'
import { createI18n } from 'vue-i18n'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
  },
  'ru-RU': {
    ...ruLocale,
  },
  'zh-CN': {
    ...zhLocale,
  },
}
const cookies = useCookies()

const getCurrentLocale = computed(() => {
  const cookieLanguage = cookies.get('i18n')
  if (cookieLanguage) {
    return cookieLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
