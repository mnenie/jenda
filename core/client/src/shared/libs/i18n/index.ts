import { computed } from 'vue'
import { createI18n } from 'vue-i18n'

import en from '../vee-validate/rules/en'
import ru from '../vee-validate/rules/ru'
import zh from '../vee-validate/rules/zh'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...en,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ru,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zh,
    },
  },
}

const getCurrentLocale = computed(() => {
  const storageLanguage = localStorage.getItem('i18n')
  if (storageLanguage) {
    return storageLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
