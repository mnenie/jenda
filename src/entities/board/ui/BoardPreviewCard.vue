<script setup lang="ts">
import type { UserType } from '@/entities/user/model';
import type { BoardPreview } from '../model';
import { UserAvatar } from '@/entities/user';
import { computed } from 'vue';

defineProps<{
  board: BoardPreview,
}>()

const userPosition = computed(() => {
  return (board: BoardPreview, user: UserType) => {
    return (board.users.length - 1 - +user._id!) * 12;
  };
});
</script>
<template>
  <div :class="$style.active_board">
    <div>
      <p class="text-base" style="font-weight: 500">{{ board.title }}</p>
      <span class="text-xs" style="color: var(--zinc-500)">{{ board.description }}</span>
    </div>
    <div :class="$style.bottom_part">
      <span class="text-xs" style="color: var(--zinc-500)"> Updated: May 2024</span>
      <div :class="$style.users">
        <template v-for="user in board.users" :key="user._id">
          <UserAvatar
            :class="$style.user"
            :style="{
              right: userPosition(board, user) + 'px',
              zIndex: userPosition(board, user),
              border: `1px solid ${user._id === '0' ? 'var(--purple-main)' : 'var(--zinc-300)'}`
            }"
          >
            <img :src="user.photoUrl" style="width: 100%" />
          </UserAvatar>
        </template>
      </div>
    </div>
  </div>
</template>

<style module lang="scss">
.active_board {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: white;
  border: 1px solid var(--zinc-200);
  min-width: 240px;
  height: 120px;
  border-radius: 15px;
  padding: 15px;
  cursor: pointer;

  .bottom_part {
    display: flex;
    align-items: flex-end;
    justify-content: space-between;
    width: 100%;

    .users {
      position: relative;
      display: flex;
      align-items: center;
      height: 20px;
      width: 20px;

      .user {
        height: 22px;
        width: 22px;
        position: absolute;
        top: 0;
      }
    }
  }
}
</style>
