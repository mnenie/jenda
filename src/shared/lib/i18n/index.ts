import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
