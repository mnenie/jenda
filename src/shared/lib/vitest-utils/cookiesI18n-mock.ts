import { vi } from 'vitest';

vi.mock('@vueuse/integrations/useCookies', () => {
  return {
    useCookies: () => ({
      get(key: string) {
        return key === 'i18n' ? 'en-US' : undefined;
      }
    })
  };
});
