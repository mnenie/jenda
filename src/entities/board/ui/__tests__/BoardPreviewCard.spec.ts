import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import '@/shared/lib/vitest-utils/cookiesI18n-mock';
import i18n from '@/shared/lib/i18n';
import BoardPreviewCard from '../BoardPreviewCard.vue';

const mockRouter = {
  push: vi.fn(),
  beforeEach: vi.fn()
};

describe('tests for BoardPreviewCard.vue', () => {
  const wrapper = mount(BoardPreviewCard, {
    global: {
      plugins: [i18n],

      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'boards.card.date_updated': 'Date Updated (test)'
          };
          return translations[key];
        },
        $router: mockRouter
      }
    },
    props: {
      board: {
        _id: '0',
        name: 'test title',
        description: 'test description',
        users: []
      }
    }
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should correctly works with i18n', () => {
    const date = wrapper.find('.bottom_part');
    expect(date.text()).toContain('Date Updated (test)');
  });

  it('should redirect to "/board/1" ', async () => {
    await wrapper.find('.active_board').trigger('click');
    expect(mockRouter.push).toHaveBeenCalledTimes(1);
    expect(mockRouter.push).toHaveBeenCalledWith('/board/1');
  });
});
