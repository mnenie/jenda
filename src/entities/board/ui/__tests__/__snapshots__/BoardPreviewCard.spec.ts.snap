// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for BoardPreviewCard > should render correctly 1`] = `
"<div class="active_board">
  <div class="text">
    <p class="text-base" style="font-weight: 500;">test title</p><span class="text-xs">test description</span>
  </div>
  <div class="bottom_part"><span class="text-xs">Date Updated (test): May 2024</span>
    <div class="users"></div>
  </div>
</div>"
`;

exports[`tests for BoardPreviewCard.vue > should render correctly 1`] = `
"<div class="active_board">
  <div class="text">
    <p class="text-base" style="font-weight: 500;">test title</p><span class="text-xs">test description</span>
  </div>
  <div class="bottom_part"><span class="text-xs">Date Updated (test): May 2024</span>
    <div class="users"></div>
  </div>
</div>"
`;
